Enable logging with Harvester upgrades

starbops opened the pull request:
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
<!-- Explain the problem you are aiming to resolve in this PR. -->

Upgrade-related logs are hard to keep:

- The logs of the upgrade Jobs are gone almost immediately after the Jobs completed, e.g., pre-drain Jobs
- Existing services like harvester, rancher, and longhorn-manager could have crucial information which might be lost during the upgrade because Pods/containers restart

And support bundles have limitations:

- It can only snapshot a point in time of the cluster; what was lost was lost.

**Solution:**
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->

Introducing UpgradeLog CRD and its accompanying controller collaborate with the existing upgrade procedure to act like a recording assistant.

For detailed information, please refer to the HEP #3274 

**Related Issue:**

#1926

**Test plan:**
<!-- Make sure tests pass on the Circle CI. -->

1. Prepare a Harvester cluster in master-head version (including this PR)
2. Upgrade with the same ISO
3. Check if the corresponding UpgradeLog created
4. After the DownloadReady condition of the UpgradeLog becomes True, try to generate an archive by executing `curl -k -X POST -u "${CATTLE_ACCESS_KEY}:${CATTLE_SECRET_KEY}" https://<harvester-vip>/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/<upgradelog-name>?action=generate`. Copy the response in stdout (it's the name of the to-be-generated archive)
5. Wait until the archive becomes ready (check this via `upgradelog.status.archives.<archiveName>.ready`), and download it by executing `curl -kfL -u "${CATTLE_ACCESS_KEY}:${CATTLE_SECRET_KEY}" https://<harvester-vip>/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/<upgradelog-name>/download?archiveName=<archiveName> -o <archiveName>.tar.gz`. Check the content of the archive.
6. Users can generate several archives per UpgradeLog and download each of them by doing step 4 and 5 repeatedly during an upgrade
7. After the upgrade is ended (either in success or failure), users can still generate archives and download them. But the content of the archives will be the same because the collection has stopped.
9. Remove the Upgrade resource should purge all the related resources, i.e., the UpgradeLog itself, the ManagedChart (if installed), the Logging resource, the ClusterFlow/Output resources, the log downloader Deployment, the log packager Jobs, and the PVC

Example of commands used to test this PR:

```
# generate an archive
$ curl -k -X POST -u token-jdpkd:k56qxnt28h4nmjp7cpqhzzgkflpkqhbhpppzmkk5bdfmr4p5j2gsn8 \
    https://192.168.48.130/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/hvst-upgrade-lz2q7-upgradelog?action=generate
"hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z" # this is the output

# download the archive
$ curl -fsSL -k -u token-jdpkd:k56qxnt28h4nmjp7cpqhzzgkflpkqhbhpppzmkk5bdfmr4p5j2gsn8 \
    https://192.168.48.130/v1/harvester/harvesterhci.io.upgradelogs/harvester-system/hvst-upgrade-lz2q7-upgradelog/download?archiveName=hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z \
    -o hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z.tar.gz

# check tarball contents
$ tar -tf hvst-upgrade-lz2q7-upgradelog-archive-2023-03-01T07-26-15Z.tar.gz
./
./logs/
./logs/kube-system.harvester-whereabouts-h6wm4.whereabouts.20230228.log
./logs/kube-system.snapshot-controller-6cc58d858f-9zrvl.snapshot-controller.20230228.log
./logs/kube-system.snapshot-controller-6cc58d858f-8mt8m.snapshot-controller.20230228.log
./logs/kube-system.kube-controller-manager-node-0.kube-controller-manager.20230228.log
./logs/longhorn-system.instance-manager-r-299489e1.replica-manager.20230228.log
./logs/kube-system.rke2-ingress-nginx-controller-5cnzf.rke2-ingress-nginx-controller.20230228.log
./logs/kube-system.kube-apiserver-node-0.kube-apiserver.20230228.log
./logs/cattle-system.rancher-5b8d85d4bc-ksfdg.rancher.20230228.log
./logs/longhorn-system.longhorn-manager-kvnlw.longhorn-manager.20230228.log
./logs/kube-system.rke2-canal-d6lqd.calico-node.20230228.log
./logs/kube-system.etcd-node-0.etcd.20230228.log
./logs/harvester-system.harvester-bcd5f8566-99qr9.apiserver.20230228.log
./logs/cattle-system.apply-hvst-upgrade-lz2q7-prepare-on-node-0-with-4a96a36ff-nfxsx.upgrade.20230228.log
./logs/longhorn-system.instance-manager-e-2db66e15.engine-manager.20230228.log
./logs/harvester-system.hvst-upgrade-lz2q7-apply-manifests-gjbgf.apply.20230228.log
```

starbops committed:
feat(upgrade): enhance upgrade observability

- Remove UpgradeLog short name due to conflicts
- Introduce 2 new UpgradeLog conditions as stages
  - UpgradeEnded for tearing down the logging infra
  - DownloadReady for indicating log archive is ready for download
- Skip missing upgradelog while cleanup

Signed-off-by: Zespre Chang <zepre.chang@suse.com>

starbops committed:
feat(upgrade): add archive gen/dl features

- New conditions and archives fields for upgradelogs holding archive
  records
- API handler for generating and downloading archives
- Add test cases
- Remove corresponding pvc while deleting upgradeLog
- Differentiate tearing down log infra and purging log data
- Make UpgradeEnded and DownloadReady conditions orthogonal
- Remove UpgradeLog status in upgrade CRD
- Set background delete for jobs
- Reduce unnecessary calls to kube-apiserver
- Add Pod affinity for log packager Job
- Decouple DownloadReady condition and each archive ready status
- Fix log downloader creation
- Tidy up log entryies
- Specify packager/downloader image while creating Jobs
- Add pod affinity for log downloader
- Turn log downloader's podAffinity to soft

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

w13915984028 reviewed (commented):
General question:

Ever consider to use `helm` to deploy the upgrade logging operator, flow and output, or use `managedchart` or `addon`, those framework will help to deploy the related parts ?  thanks.

starbops committed:
Implement operator deployment phase

- Detect existing rancher-logging Addon or ManagedChart
- Install one if none of the above exists
- Set OperatorReady to true when the ManagedChart is ready
- Remove the ManagedChart when upgrade log is stopped
- Use rancher mirrored images
- New status field for Upgrade CRD: UpgradeLog
- Update UpgradeLog remove path
- Purge when Upgrade resource was deleted
- Purge when "read-message" was labeled on Upgrade resource
- Move the logging-operator to cattle-logging-system

This will separate the controlNamespaces specified by the
UpgradeLog-installed Logging resource and the rancher-logging-installed
Logging resource. The rancher-logging config-check Pod will not read the
unexpected config.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

FrankYang0529 reviewed (approved):
Test and LGTM. Thanks.

starbops committed:
Update APIs to align with other resources

- Add "generate" archive action
- Add "download" link with archive name as a query string
- Change UpgradeLog.Spec.Upgrade to UpgradeName
- Add retry for get archive http requests
- Fix log downloader endpoint with a Service object
- Especially for error code 4, this is due to log line incomplete, jq
  considers the line is broken therefore stop processing.
- Change compression format to zip to align with SB
- Tear down UpgradeLog when the Upgrade is missing

Normally, the UpgradeLog CR will be deleted when the corresponding
Upgrade CR is removed. But this provides us a double insurance to
clean the mess up.

Note: this commit only addresses the situation that happened at the
beginning of the UpgradeLogReady phase. No further checks after the
point in time.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

bk201 reviewed (approved):
LGTM!

w13915984028 reviewed (commented):
A few questions:

w13915984028 reviewed (commented):
At first, thanks @starbops has done a lot of work to implement this feature from beginning.

The `pkg/controller/master/upgradelog/controller.go` has an internal state machine to control the deployment of log related mechanism, and the overall feature is embedded into `upgrade` process.

I am afraid we can test all throughly in v1.1.2.

In case there are bugs happening inbetween the upgrade, do we have quick way to recover from the middle state of it?

guangbochen reviewed (approved):
LGTM, please help to squish the commits before merge, thanks.

w13915984028 reviewed (approved):
LGTM, thanks for your great effort @starbops .

bk201 reviewed (approved):
Thanks for the great feature.
